neqo-transport/src/connection/tests/stream.rs

```diff
@@ -111,6 +113,121 @@ fn transfer() {
+    connect_force_idle(&mut client, &mut server);
+
+    qdebug!("---- client sends");
+    // open all streams and set the sendorders
+    let mut ordered = BinaryHeap::new();
+    let mut streams = Vec::<StreamId>::new();
+    for sendorder in order_of_sendorder {
+	streams.push(client.stream_create(StreamType::UniDi).unwrap());
+	let id = *(streams.last()).unwrap();
+	ordered.push(IdEntry { sendorder: *sendorder, stream_id: id });
+	client.streams.set_sendorder(id, sendorder.sendorder).ok();
+    }
+    // Write some data to all the streams
+    for stream_id in streams {
+	client.stream_send(stream_id, &[6; 100]).unwrap();
+    }
+
+    // Sending this much takes a few datagrams.
+    let mut datagrams = vec![];
+    let mut out = client.process_output(now());
+    while let Some(d) = out.dgram() {
+        datagrams.push(d);
+        out = client.process_output(now());
+    }
+    assert_eq!(*client.state(), State::Confirmed);
+
+    qdebug!("---- server receives");
+    for (d_num, d) in datagrams.into_iter().enumerate() {
+        let out = server.process(Some(d), now());
+        assert_eq!(
+            out.as_dgram_ref().is_some(),
+            (d_num + 1) % usize::try_from(DEFAULT_ACK_PACKET_TOLERANCE + 1).unwrap() == 0
+        );
+        qdebug!("Output={:0x?}", out.as_dgram_ref());
+    }
+    assert_eq!(*server.state(), State::Confirmed);
+
+    let mut stream_ids = server.events().filter_map(|evt| match evt {
+        ConnectionEvent::RecvStreamReadable { stream_id, .. } => Some(stream_id),
+        _ => None,
+    });
+
+    // streams should arrive in priority order, not order of creation, if sendorder prioritization
+    // is working correctly
+    let mut recv_streams = Vec::<StreamId>::new();
+    let mut should_recv_streams =  Vec::<StreamId>::new();
+    // walk the ordered list of streams, and handle random ordering within a sendorder
+    let mut last_sendorder: StreamOrder = StreamOrder{sendorder: None};
+    // BinaryHeap's sorted iterator is nightly only for now
+    // the ordering is inverted, however!
+    let mut ordered_vec = ordered.into_sorted_vec();
+    ordered_vec.reverse();
+    for identry in ordered_vec.iter() {
+	let sendorder = identry.sendorder;
+	let stream_id = identry.stream_id;
+	// collect set of streams at the same priority
+	if sendorder != last_sendorder {
+	    // do checks on the list received
+	    for id in &should_recv_streams {
+		// check if this is the right id
+		assert!(recv_streams.contains(&id));
+	    }
+	    recv_streams.clear();
+	    should_recv_streams.clear();
+	    last_sendorder = sendorder
+	}
+	recv_streams.push(stream_ids.next().expect("should have data on a new stream"));
+	should_recv_streams.push(stream_id);
+    }
+    // Check the last bucket
+    for id in &should_recv_streams {
+	// check if this is the right id
+	assert!(recv_streams.contains(&id));
+    }
+
+    assert!(stream_ids.next().is_none());
+}
+
+macro_rules! sendorder_tests {
```

Comment:
I have an allergy to macros that are just functions.  You should be able to do something like this:

```rust
#[test]
// TODO pick a nicer name
fn sendorder_0() {
  sendorder_test(&[None, Some(1), Some(2), Some(3)]);
}
```

The overall shape isn't that much different (other than the test function taking `&[Option<SendOrder>]` rather than a `Vec`), but you don't have a macro.